pkg/registry/storage/csidriver/strategy.go

```diff
@@ -0,0 +1,111 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csidriver
+
+import (
+	"context"
+
+	storageapiv1beta1 "k8s.io/api/storage/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/apis/storage"
+	"k8s.io/kubernetes/pkg/apis/storage/validation"
+)
+
+// csiDriverStrategy implements behavior for CSIDriver objects
+type csiDriverStrategy struct {
+	runtime.ObjectTyper
+	names.NameGenerator
+}
+
+// Strategy is the default logic that applies when creating and updating
+// CSIDriver objects via the REST API.
+var Strategy = csiDriverStrategy{legacyscheme.Scheme, names.SimpleNameGenerator}
+
+func (csiDriverStrategy) NamespaceScoped() bool {
+	return false
+}
+
+// ResetBeforeCreate clears the Status field which is not allowed to be set by end users on creation.
+func (csiDriverStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
+}
+
+func (csiDriverStrategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
+	csiDriver := obj.(*storage.CSIDriver)
+
+	errs := validation.ValidateCSIDriver(csiDriver)
+
+	var groupVersion schema.GroupVersion
+
+	if requestInfo, found := genericapirequest.RequestInfoFrom(ctx); found {
+		groupVersion = schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}
+	}
+
+	switch groupVersion {
+	case storageapiv1beta1.SchemeGroupVersion:
+		errs = append(errs, validation.ValidateCSIDriver(csiDriver)...)
+	default:
+		// no extra validation
+	}
+	return errs
+}
+
+// Canonicalize normalizes the object after validation.
+func (csiDriverStrategy) Canonicalize(obj runtime.Object) {
+}
+
+func (csiDriverStrategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+// PrepareForUpdate sets the Status fields which is not allowed to be set by an end user updating a CSIDriver
+func (csiDriverStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+}
+
+func (csiDriverStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
+	newCSIDriverObj := obj.(*storage.CSIDriver)
+	oldCSIDriverObj := old.(*storage.CSIDriver)
+	errorList := validation.ValidateCSIDriver(newCSIDriverObj)
+	return append(errorList, validation.ValidateCSIDriverUpdate(newCSIDriverObj, oldCSIDriverObj)...)
```

Comment:
not sure if this is just a common pattern i'm not familiar with but it feels weird to me to validate the old object separately, then do a "update validation" (which validates the new driver inside it).

Can `ValidateCSIDriverUpdate` just validate both old and new? Or neither, and leave the single object validation to the caller?

Again, not sure if this is an established pattern, feel free to correct me.

Reply:
I'm also not familiar with this.  I followed this for VolumeAttachment: https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/pkg/registry/storage/volumeattachment/strategy.go#L116

I guess it first makes sure the newObj is okay, then validates update works.

Reply:
There's usually two steps when validating an update:
1. Check that the updated (new) object passes regular object validation
2. Check that changes between old and new are allowed